File: frontend/src/services/payroll/payrollService.js
// ==================== PAYROLL SERVICE ====================
// frontend/src/services/payroll/payrollService.js
import apiClient from '../../utils/apiClient.js';

const PAYROLL_ENDPOINTS = {
    BASE: '/api/v1/payroll',
    GENERATE_MONTHLY: (year, month) => `/api/v1/payroll/generate-monthly/${year}/${month}`,
    GENERATE_MONTHLY_PARAM: '/api/v1/payroll/generate-monthly',
    REPORT: '/api/v1/payroll/report',
    SUMMARY: '/api/v1/payroll/summary',
    EMPLOYEE: (employeeId) => `/api/v1/payroll/employee/${employeeId}`,
    PERIOD: '/api/v1/payroll/period',
    FINALIZE_PERIOD: '/api/v1/payroll/finalize-period',
    SEND_PERIOD: '/api/v1/payroll/send-period',
    STATUS: (status) => `/api/v1/payroll/status/${status}`,
    STATUS_SUMMARY: '/api/v1/payroll/status-summary',
    EXISTS: (year, month) => `/api/v1/payroll/exists/${year}/${month}`,
    MISSING_PAYSLIPS: (year, month) => `/api/v1/payroll/missing-payslips/${year}/${month}`,
    GENERATE_SPECIFIC: '/api/v1/payroll/generate-specific',
    MONTHLY_STATS: (year) => `/api/v1/payroll/monthly-stats/${year}`,
    VALIDATE: (year, month) => `/api/v1/payroll/validate/${year}/${month}`,
    REPROCESS_FAILED: '/api/v1/payroll/reprocess-failed',
    EXPORT: '/api/v1/payroll/export'
};

export const payrollService = {
    // Generate monthly payslips for all employees
    generateMonthlyPayslips: (year, month, createdBy = 'SYSTEM') => {
        return apiClient.post(PAYROLL_ENDPOINTS.GENERATE_MONTHLY(year, month), null, {
            params: { createdBy }
        });
    },

    // Generate monthly payslips with YearMonth parameter
    generateMonthlyPayslipsWithParam: (payPeriod, createdBy = 'SYSTEM') => {
        return apiClient.post(PAYROLL_ENDPOINTS.GENERATE_MONTHLY_PARAM, null, {
            params: { payPeriod, createdBy }
        });
    },

    // Get payroll report for a period
    getPayrollReport: (startDate, endDate) => {
        return apiClient.get(PAYROLL_ENDPOINTS.REPORT, {
            params: { startDate, endDate }
        });
    },

    // Get payroll summary (more efficient than full report)
    getPayrollSummary: (startDate, endDate) => {
        return apiClient.get(PAYROLL_ENDPOINTS.SUMMARY, {
            params: { startDate, endDate }
        });
    },

    // Get payslips by employee
    getPayslipsByEmployee: (employeeId) => {
        return apiClient.get(PAYROLL_ENDPOINTS.EMPLOYEE(employeeId));
    },

    // Get payslips by period
    getPayslipsByPeriod: (startDate, endDate) => {
        return apiClient.get(PAYROLL_ENDPOINTS.PERIOD, {
            params: { startDate, endDate }
        });
    },

    // Bulk finalize payslips for a period
    finalizePayslipsForPeriod: (startDate, endDate, approvedBy = 'SYSTEM') => {
        return apiClient.post(PAYROLL_ENDPOINTS.FINALIZE_PERIOD, null, {
            params: { startDate, endDate, approvedBy }
        });
    },

    // Bulk send payslips for a period
    sendPayslipsForPeriod: (startDate, endDate) => {
        return apiClient.post(PAYROLL_ENDPOINTS.SEND_PERIOD, null, {
            params: { startDate, endDate }
        });
    },

    // Get payslips by status for a period
    getPayslipsByStatusAndPeriod: (status, startDate, endDate) => {
        return apiClient.get(PAYROLL_ENDPOINTS.STATUS(status), {
            params: { startDate, endDate }
        });
    },

    // Get payroll status summary
    getPayrollStatusSummary: (startDate, endDate) => {
        return apiClient.get(PAYROLL_ENDPOINTS.STATUS_SUMMARY, {
            params: { startDate, endDate }
        });
    },

    // Check if payroll exists for a period
    payrollExistsForPeriod: (year, month) => {
        return apiClient.get(PAYROLL_ENDPOINTS.EXISTS(year, month));
    },

    // Get employees without payslips for a period
    getEmployeesWithoutPayslipsForPeriod: (year, month) => {
        return apiClient.get(PAYROLL_ENDPOINTS.MISSING_PAYSLIPS(year, month));
    },

    // Generate payslips for specific employees only
    generatePayslipsForEmployees: (employeeIds, payPeriod, createdBy = 'SYSTEM') => {
        return apiClient.post(PAYROLL_ENDPOINTS.GENERATE_SPECIFIC, employeeIds, {
            params: { payPeriod, createdBy }
        });
    },

    // Get monthly payroll statistics
    getMonthlyPayrollStats: (year) => {
        return apiClient.get(PAYROLL_ENDPOINTS.MONTHLY_STATS(year));
    },

    // Validate payroll data for a period
    validatePayrollForPeriod: (year, month) => {
        return apiClient.post(PAYROLL_ENDPOINTS.VALIDATE(year, month));
    },

    // Reprocess failed payslips
    reprocessFailedPayslips: (startDate, endDate, reprocessedBy = 'SYSTEM') => {
        return apiClient.post(PAYROLL_ENDPOINTS.REPROCESS_FAILED, null, {
            params: { startDate, endDate, reprocessedBy }
        });
    },

    // Export payroll data
    exportPayrollData: (startDate, endDate, format = 'excel') => {
        return apiClient.get(PAYROLL_ENDPOINTS.EXPORT, {
            params: { startDate, endDate, format },
            responseType: 'blob'
        });
    }
};

File: backend/src/main/java/com/example/backend/services/payroll/PayrollService.java
package com.example.backend.services.payroll;

import com.example.backend.dto.payroll.PayrollReportDTO;
import com.example.backend.dto.payroll.PayslipDTO;
import com.example.backend.models.hr.Employee;
import com.example.backend.models.payroll.Payslip;
import com.example.backend.repositories.hr.EmployeeRepository;
import com.example.backend.repositories.payroll.PayslipRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class PayrollService {

    private final PayslipRepository payslipRepository;
    private final EmployeeRepository employeeRepository;
    private final PayslipService payslipService;

    /**
     * Generate payslips for all active employees for a given period
     */
    @Transactional
    public List<PayslipDTO> generateMonthlyPayslips(YearMonth payPeriod, String createdBy) {
        log.info("Generating monthly payslips for period: {}", payPeriod);

        LocalDate payPeriodStart = payPeriod.atDay(1);
        LocalDate payPeriodEnd = payPeriod.atEndOfMonth();
        LocalDate payDate = payPeriodEnd.plusDays(1); // Pay on first day of next month

        // Get LIST of active employees
        List<Employee> activeEmployees = employeeRepository.findByStatus("ACTIVE");

        if (activeEmployees.isEmpty()) {
            log.warn("No active employees found for payroll generation");
            return List.of(); // Return empty list instead of null
        }

        log.info("Found {} active employees for payroll generation", activeEmployees.size());

        return activeEmployees.stream()
                .map(employee -> {
                    try {
                        // Check if payslip already exists
                        if (payslipRepository.existsByEmployeeIdAndPayPeriodStartAndPayPeriodEnd(
                                employee.getId(), payPeriodStart, payPeriodEnd)) {
                            log.warn("Payslip already exists for employee {} for period {}",
                                    employee.getId(), payPeriod);
                            return null;
                        }

                        log.info("Generating payslip for employee: {} {}",
                                employee.getFirstName(), employee.getLastName());

                        return payslipService.generatePayslip(employee, payPeriodStart, payPeriodEnd, payDate, createdBy);
                    } catch (Exception e) {
                        log.error("Error generating payslip for employee {} ({}): {}",
                                employee.getId(), employee.getFirstName() + " " + employee.getLastName(), e.getMessage());
                        // Don't throw here - continue processing other employees
                        return null;
                    }
                })
                .filter(payslip -> payslip != null)
                .collect(Collectors.toList());
    }

    /**
     * Get payroll report for a specific period
     */
    public PayrollReportDTO getPayrollReport(LocalDate startDate, LocalDate endDate) {
        log.info("Generating payroll report for period: {} to {}", startDate, endDate);

        List<Payslip> payslips = payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate);

        BigDecimal totalGrossPayroll = payslips.stream()
                .map(Payslip::getGrossSalary)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalNetPayroll = payslips.stream()
                .map(Payslip::getNetPay)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalDeductions = payslips.stream()
                .map(Payslip::getTotalDeductions)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalEmployerContributions = payslips.stream()
                .map(Payslip::getTotalEmployerContributions)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return PayrollReportDTO.builder()
                .reportPeriodStart(startDate)
                .reportPeriodEnd(endDate)
                .totalEmployees(payslips.size())
                .totalGrossPayroll(totalGrossPayroll)
                .totalNetPayroll(totalNetPayroll)
                .totalDeductions(totalDeductions)
                .totalEmployerContributions(totalEmployerContributions)
                .build();
    }

    /**
     * Get payslips by employee
     */
    public List<PayslipDTO> getPayslipsByEmployee(UUID employeeId) {
        log.info("Getting payslips for employee: {}", employeeId);
        return payslipRepository.findByEmployeeIdOrderByPayDateDesc(employeeId)
                .stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get payslips by period
     */
    public List<PayslipDTO> getPayslipsByPeriod(LocalDate startDate, LocalDate endDate) {
        log.info("Getting payslips for period: {} to {}", startDate, endDate);
        return payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate)
                .stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Bulk finalize payslips for a period
     */
    @Transactional
    public List<PayslipDTO> finalizePayslipsForPeriod(LocalDate startDate, LocalDate endDate, String approvedBy) {
        log.info("Finalizing payslips for period: {} to {}", startDate, endDate);

        List<Payslip> draftPayslips = payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate)
                .stream()
                .filter(p -> p.getStatus() == Payslip.PayslipStatus.DRAFT)
                .collect(Collectors.toList());

        if (draftPayslips.isEmpty()) {
            log.warn("No draft payslips found for period: {} to {}", startDate, endDate);
            return List.of();
        }

        log.info("Found {} draft payslips to finalize", draftPayslips.size());

        return draftPayslips.stream()
                .map(payslip -> {
                    try {
                        return payslipService.finalizePayslip(payslip.getId(), approvedBy);
                    } catch (Exception e) {
                        log.error("Error finalizing payslip {} for employee {}: {}",
                                payslip.getId(), payslip.getEmployee().getFullName(), e.getMessage());
                        return null;
                    }
                })
                .filter(payslip -> payslip != null)
                .collect(Collectors.toList());
    }

    /**
     * Bulk send payslips for a period
     */
    @Transactional
    public List<PayslipDTO> sendPayslipsForPeriod(LocalDate startDate, LocalDate endDate) {
        log.info("Sending payslips for period: {} to {}", startDate, endDate);

        List<Payslip> approvedPayslips = payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate)
                .stream()
                .filter(p -> p.getStatus() == Payslip.PayslipStatus.APPROVED)
                .collect(Collectors.toList());

        if (approvedPayslips.isEmpty()) {
            log.warn("No approved payslips found for period: {} to {}", startDate, endDate);
            return List.of();
        }

        log.info("Found {} approved payslips to send", approvedPayslips.size());

        return approvedPayslips.stream()
                .map(payslip -> {
                    try {
                        return payslipService.sendPayslip(payslip.getId());
                    } catch (Exception e) {
                        log.error("Error sending payslip {} for employee {}: {}",
                                payslip.getId(), payslip.getEmployee().getFullName(), e.getMessage());
                        return null;
                    }
                })
                .filter(payslip -> payslip != null)
                .collect(Collectors.toList());
    }

    /**
     * Get payroll summary statistics
     */
    public PayrollReportDTO getPayrollSummary(LocalDate startDate, LocalDate endDate) {
        log.info("Getting payroll summary for period: {} to {}", startDate, endDate);

        // Get totals using repository aggregation methods (more efficient)
        BigDecimal totalGrossPayroll = payslipRepository.getTotalGrossPayrollByPeriod(startDate, endDate);
        BigDecimal totalNetPayroll = payslipRepository.getTotalNetPayrollByPeriod(startDate, endDate);
        BigDecimal totalDeductions = payslipRepository.getTotalDeductionsByPeriod(startDate, endDate);

        // Count payslips
        long totalPayslips = payslipRepository.countByPayPeriodStartBetween(startDate, endDate);

        // Handle null values from aggregation
        totalGrossPayroll = totalGrossPayroll != null ? totalGrossPayroll : BigDecimal.ZERO;
        totalNetPayroll = totalNetPayroll != null ? totalNetPayroll : BigDecimal.ZERO;
        totalDeductions = totalDeductions != null ? totalDeductions : BigDecimal.ZERO;

        return PayrollReportDTO.builder()
                .reportPeriodStart(startDate)
                .reportPeriodEnd(endDate)
                .totalEmployees((int) totalPayslips)
                .totalGrossPayroll(totalGrossPayroll)
                .totalNetPayroll(totalNetPayroll)
                .totalDeductions(totalDeductions)
                .totalEmployerContributions(BigDecimal.ZERO) // Calculate this if needed
                .build();
    }

    /**
     * Get payslips by status for a period
     */
    public List<PayslipDTO> getPayslipsByStatusAndPeriod(Payslip.PayslipStatus status,
                                                         LocalDate startDate, LocalDate endDate) {
        log.info("Getting payslips with status {} for period: {} to {}", status, startDate, endDate);

        return payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate)
                .stream()
                .filter(p -> p.getStatus() == status)
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get employee count by payroll status for a period
     */
    public Map<String, Long> getPayrollStatusSummary(LocalDate startDate, LocalDate endDate) {
        log.info("Getting payroll status summary for period: {} to {}", startDate, endDate);

        List<Payslip> payslips = payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate);

        Map<String, Long> statusCounts = payslips.stream()
                .collect(Collectors.groupingBy(
                        p -> p.getStatus().name(),
                        Collectors.counting()
                ));

        // Ensure all statuses are represented
        for (Payslip.PayslipStatus status : Payslip.PayslipStatus.values()) {
            statusCounts.putIfAbsent(status.name(), 0L);
        }

        return statusCounts;
    }

    /**
     * Check if payroll exists for period
     */
    public boolean payrollExistsForPeriod(YearMonth payPeriod) {
        LocalDate startDate = payPeriod.atDay(1);
        LocalDate endDate = payPeriod.atEndOfMonth();

        return payslipRepository.countByPayPeriodStartBetween(startDate, endDate) > 0;
    }

    /**
     * Get employees without payslips for a period
     */
    public List<Employee> getEmployeesWithoutPayslipsForPeriod(YearMonth payPeriod) {
        log.info("Finding employees without payslips for period: {}", payPeriod);

        LocalDate startDate = payPeriod.atDay(1);
        LocalDate endDate = payPeriod.atEndOfMonth();

        List<Employee> activeEmployees = employeeRepository.findByStatus("ACTIVE");
        List<Payslip> existingPayslips = payslipRepository.findByPayPeriodStartBetweenOrderByPayDateDesc(startDate, endDate);

        List<UUID> employeesWithPayslips = existingPayslips.stream()
                .map(p -> p.getEmployee().getId())
                .collect(Collectors.toList());

        return activeEmployees.stream()
                .filter(e -> !employeesWithPayslips.contains(e.getId()))
                .collect(Collectors.toList());
    }

    /**
     * Enhanced convert to DTO method to match PayslipService
     */
    private PayslipDTO convertToDTO(Payslip payslip) {
        return PayslipDTO.builder()
                .id(payslip.getId())
                .employeeId(payslip.getEmployee().getId())
                .employeeName(payslip.getEmployee().getFullName()) // Use getFullName() method
                .employeeEmail(payslip.getEmployee().getEmail())
                .jobPositionName(payslip.getEmployee().getJobPosition() != null ?
                        payslip.getEmployee().getJobPosition().getPositionName() : "")
                .departmentName(payslip.getEmployee().getJobPosition() != null &&
                        payslip.getEmployee().getJobPosition().getDepartment() != null ?
                        payslip.getEmployee().getJobPosition().getDepartment().getName() : "")
                .payPeriodStart(payslip.getPayPeriodStart())
                .payPeriodEnd(payslip.getPayPeriodEnd())
                .payDate(payslip.getPayDate())
                .grossSalary(payslip.getGrossSalary())
                .netPay(payslip.getNetPay())
                .totalEarnings(payslip.getTotalEarnings())
                .totalDeductions(payslip.getTotalDeductions())
                .totalEmployerContributions(payslip.getTotalEmployerContributions())
                .daysWorked(payslip.getDaysWorked())
                .daysAbsent(payslip.getDaysAbsent())
                .overtimeHours(payslip.getOvertimeHours())
                .status(payslip.getStatus().name())
                .pdfPath(payslip.getPdfPath())
                .generatedAt(payslip.getGeneratedAt())
                .sentAt(payslip.getSentAt())
                .acknowledgedAt(payslip.getAcknowledgedAt())
                .createdBy(payslip.getCreatedBy())
                .build();
    }
}